cunumeric/deferred.py

```diff
@@ -421,14 +421,28 @@ def _create_indexing_array(self, key, is_set=False):
         rhs = self
         # the index where the first index_array is passed to the [] operator
         start_index = -1
-        if (
-            isinstance(key, NumPyThunk)
-            and key.dtype == bool
-            and key.shape == rhs.shape
-        ):
+        if isinstance(key, NumPyThunk) and key.dtype == bool:
             if not isinstance(key, DeferredArray):
                 key = self.runtime.to_deferred_array(key)
 
+            # in case when boolean array is passed as an index, shape for all
+            # its dimensions should be the same as the shape of
+            # corresponding dimensions of the input array
+            for i in range(key.ndim):
+                if key.shape[i] != rhs.shape[i]:
+                    raise ValueError(
+                        "shape of the index array for "
+                        f"dimension {i} doesn't match to the shape of the"
+                        f"index array which is {rhs.shape[i]}"
+                    )
+            key_dims = key.ndim
+            out_dim = rhs.ndim - key_dims + 1
+            key = key.base
```

Comment:
I know this is legal in Python, but assigning values of two different types to the same variable is an anti-pattern. This is also problematic because we will (actually bryan will) eventually put type annotations on the cunumeric code and this line introduces a rather unnecessary union type.